src/plot.js

```diff
@@ -486,73 +530,133 @@ function nolabel(axis) {
+// isn’t sorted because it’s constructed prior to the other mark channels.)
+function facetKeys(facets, fx, fy) {
+  const fxI = fx && new InternMap(fx.domain().map((x, i) => [x, i]));
+  const fyI = fy && new InternMap(fy.domain().map((y, i) => [y, i]));
+  return sort(facets, (a, b) => (fxI && fxI.get(a.x) - fxI.get(b.x)) || (fyI && fyI.get(a.y) - fyI.get(b.y)));
+}
+
+// Returns a (possibly nested) Map of [[key1, index1], [key2, index2], …]
+// representing the data indexes associated with each facet.
+function facetGroups(data, fx, fy) {
+  const index = range(data);
   return fx && fy ? facetGroup2(index, fx, fy) : fx ? facetGroup1(index, fx) : facetGroup1(index, fy);
 }
 
 function facetGroup1(index, {value: F}) {
-  return groups(index, (i) => F[i]);
+  return group(index, (i) => F[i]);
 }
 
 function facetGroup2(index, {value: FX}, {value: FY}) {
-  return groups(
+  return group(
     index,
     (i) => FX[i],
     (i) => FY[i]
-  ).flatMap(([x, xgroup]) => xgroup.map(([y, ygroup]) => [[x, y], ygroup]));
+  );
 }
 
-// This must match the key structure returned by facetGroups.
 function facetTranslate(fx, fy) {
   return fx && fy
-    ? ([kx, ky]) => `translate(${fx(kx)},${fy(ky)})`
+    ? ({x, y}) => `translate(${fx(x)},${fy(y)})`
     : fx
-    ? (kx) => `translate(${fx(kx)},0)`
-    : (ky) => `translate(0,${fy(ky)})`;
+    ? ({x}) => `translate(${fx(x)},0)`
+    : ({y}) => `translate(0,${fy(y)})`;
 }
 
-function facetMap({fx, fy}) {
-  return new (fx && fy ? FacetMap2 : FacetMap)();
+// Returns an index that for each facet lists all the elements present in other
+// facets in the original index.
+function excludeIndex(index) {
+  const ex = [];
+  const e = new Uint32Array(sum(index, (d) => d.length));
+  for (const i of index) {
+    let n = 0;
+    for (const j of index) {
+      if (i === j) continue;
+      e.set(j, n);
+      n += j.length;
+    }
+    ex.push(e.slice(0, n));
+  }
+  return ex;
 }
 
-class FacetMap {
-  constructor() {
-    this._ = new InternMap();
-  }
-  has(key) {
-    return this._.has(key);
-  }
-  get(key) {
-    return this._.get(key);
-  }
-  set(key, value) {
-    return this._.set(key, value), this;
-  }
+// Returns the facet groups, and possibly fx and fy channels, associated with
+// the top-level facet option {data, x, y}.
+function maybeTopFacet(facet) {
+  if (facet == null) return;
+  const {x, y} = facet;
+  if (x == null && y == null) return;
+  const data = arrayify(facet.data);
+  if (data == null) throw new Error(`missing facet data`);
+  // TODO Call applyScaleTransforms on the fx and fy channels.
+  const fx = x != null ? Channel(data, {value: x, scale: "fx"}) : undefined;
+  const fy = y != null ? Channel(data, {value: y, scale: "fy"}) : undefined;
+  const groups = facetGroups(data, fx, fy);
+  // If the top-level faceting is non-trivial, track the corresponding data
+  // length, in order to compare it for the warning above.
```

Comment:
What does “non-trivial” faceting mean? That there’s more than one facet?

Reply:
yes

Reply:
edited the comment